src/features/matrix_shortcuts.ts

```diff
@@ -1,59 +1,174 @@
 import { EditorView } from "@codemirror/view";
-import { setCursor } from "src/utils/editor_utils";
-import { getLatexSuiteConfig } from "src/snippets/codemirror/config";
+import { EditorSelection, SelectionRange } from "@codemirror/state";
 import { Context } from "src/utils/context";
+import { setCursor, replaceRange } from "src/utils/editor_utils";
+import { getLatexSuiteConfig } from "src/snippets/codemirror/config";
 import { tabout } from "src/features/tabout";
 
+
+const ALIGNMENT = " & ";
+const LINE_BREAK = " \\\\\n"
+const LINE_BREAK_INLINE = " \\\\ "
+const END_LINE_BREAK = LINE_BREAK.trimEnd();
+let trimWhitespace = false;
+let trimAlignment = false;
+let hlineLineBreakEnabled = false;
+let trimEmptyLineAfterEnv = false;
+let addLineBreakAfterEnv = false;
+
+
+const isLineBreak = (separator: string) => {
+	return separator.contains("\\\\");
+}
+
+
+const isMultiLineBreak = (separator: string): boolean => {
+	return separator.contains("\\\\") && separator.contains("\n");
+}
+
+
+const isHline = (line: string): boolean => {
+	return line.trimEnd().endsWith("\\hline");
+}
+
+
+const generateSeparatorChange = (separator: string, view: EditorView, range: SelectionRange): { from: number, to: number, insert: string } => {
+	const d = view.state.doc;
+
+	const fromLine = d.lineAt(range.from);
+	const textBeforeFrom = d.sliceString(fromLine.from, range.from).trimStart();  // Preserve indents
+
+	const toLine = d.lineAt(range.from);
+	const textAfterTo = d.sliceString(range.to, toLine.to);
+
+	let from = range.from;
+	let to = range.to;
```

Comment:
```suggestion
	let { from, to } = range;
```
this doesn't matter that much, but I prefer this style.